File: 3471-find-the-largest-almost-missing-integer/3471-find-the-largest-almost-missing-integer.ts
function largestInteger(nums: number[], k: number): number {
    const n = nums.length;
    const freq = new Map<number, number>();
    for (const x of nums) freq.set(x, (freq.get(x) || 0) + 1);

    if (k === n) return Math.max(...nums);

    if (k === 1) {
        let ans = -1;
        for (const x of nums) {
            if (freq.get(x) === 1) ans = Math.max(ans, x);
        }
        return ans;
    }

    const first = nums[0];
    const last = nums[n - 1];

    if (first === last) return -1;

    const firstValid = (freq.get(first) === 1);
    const lastValid  = (freq.get(last) === 1);

    if (firstValid && lastValid) return Math.max(first, last);
    if (firstValid) return first;
    if (lastValid) return last;
    return -1;
}

File: 3471-find-the-largest-almost-missing-integer/3471-find-the-largest-almost-missing-integer.rs
use std::collections::HashMap;

impl Solution {
    pub fn largest_integer(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let mut freq = HashMap::new();
        for &x in &nums {
            *freq.entry(x).or_insert(0) += 1;
        }

        if k as usize == n {
            return *nums.iter().max().unwrap();
        }

        if k == 1 {
            let mut ans = -1;
            for &x in &nums {
                if freq[&x] == 1 {
                    ans = ans.max(x);
                }
            }
            return ans;
        }

        let first = nums[0];
        let last = nums[n - 1];

        if first == last {
            return -1;
        }

        let first_valid = freq[&first] == 1;
        let last_valid = freq[&last] == 1;

        if first_valid && last_valid {
            return first.max(last);
        }
        if first_valid {
            return first;
        }
        if last_valid {
            return last;
        }
        -1
    }
}